File: leetcode/Valid_Parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character,Character> pair = new HashMap<>();
        pair.put(')','('); pair.put('}','{'); pair.put(']','[');
        Stack<Character> stack = new Stack<>();
        
        for (int i=0; i<s.length(); ++i) {
            char c = s.charAt(i);
            
            // opening bracket
            if (pair.containsValue(c)) {
                stack.push(c);
                continue;
            }
            
            // no opening bracket
            if (stack.isEmpty()) 
                return false;
            
            // closing bracket
            char opening = stack.pop();
            if (opening != pair.get(c))
                return false;
        }
        return stack.isEmpty();
    }
}

File: LeetCode/Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = {')':'(', ']':'[', '}':'{'}
        stack = []
        
        for char in s:
            if char not in mapping: # (, [, {
                stack.append(char)
            else:
                if not stack: # 처음에 ), ], } 가 들어온 경우
                    return False

                tmp = stack.pop()
                if mapping[char] != tmp:
                    return False

        if stack: # 스택이 남아있는 경우
            return False
        return True
